src/xref2/env.ml

```diff
@@ -123,7 +133,8 @@ end = struct
         let mark_shadow e =
           if e.kind = kind then { e with shadowed = true } else e
         in
-        if List.exists has_shadow tl then List.map mark_shadow tl else tl
+        if shadow && List.exists has_shadow tl then List.map mark_shadow tl
+        else tl
```

Comment:
This seems a bit awkward. I wonder if it would be cleaner to handle the shadowing on the lookup path? It seems that the 'shadowed' bool is a bit redundant - we could recalculate it even if it wasn't being stored.

Reply:
The problem with recomputing it in `find` is that it depends on the kind, which might be too heavy for the fast path.
I've changed the code to remove the shadowing bit entirely. There's now two maps, one where shadowed elements are removed and an other for identifiers.